gtsam/inference/BayesTree.h

```diff
@@ -111,6 +113,25 @@ namespace gtsam {
     /** Copy constructor */
     BayesTree(const This& other);
 
+    ~BayesTree() {
+      for (auto&& root: roots_) {
+        std::queue<Clique*> bfs_queue;
+        std::deque<Clique*> topological_order;
+        bfs_queue.push(root.get());
+        while (!bfs_queue.empty()) {
+          Clique* current = bfs_queue.front();
+          bfs_queue.pop();
+          topological_order.push_front(current);
+          for (auto&& child: current->children) {
+            bfs_queue.push(child.get());
+          }
+        }
+        for (auto&& clique: topological_order) {
+          clique->children.clear();
```

Comment:
Basically what you want is to systematically reduce the reference count by 1. I think this can be safely done by keeping a `std::vector<shared_ptr>` sorted topologically, and then set the `roots_` to `nullptr`. After that you delete this vector, which should do the same thing as this code but keeps children that should not be deleted alive.

Reply:
I see, let me try that.

Reply:
Hi @ProfFan , I adopt your idea and improved it a bit. Instead of having an extra vector, we can actually use the queue for BFS to keep the extra reference, so how it works is like:
- add root node to queue and set original root ptr to nullptr
- start BFS the queue, which will iterate through the tree by inversed topological order.
- if any node, e.g. the root node, is only referenced by the queue, when BFS processing it, its children will first be add to the queue, then it will be automatically released, while its children won't immediately as the queue still holds their reference.

I've updated this PR according to this idea. It seems the segment fault issue is also fixed with this solution. 
If such a fix is acceptable, can I open a PR to [borglab/gtsam](https://github.com/borglab/gtsam)?

Reply:
Very cool! I’ll let @ProfFan weigh in on the detail. Does this fix the subtree issue or is that orthogonal?

Reply:
I believe it does fix the subtree issue.

Reply:
Yeah go ahead :)